File: made-react-app/lib/shell-type-utils.ts
import fs from "fs";
import path from "path";
import yaml from "js-yaml";
import { getGeneratorPath } from "@/lib/aas-config";

function deepMerge(
  base: Record<string, unknown>,
  override: Record<string, unknown>
): Record<string, unknown> {
  const result: Record<string, unknown> = { ...base };
  for (const [key, val] of Object.entries(override)) {
    if (
      key in result &&
      typeof result[key] === "object" && !Array.isArray(result[key]) && result[key] !== null &&
      typeof val === "object" && !Array.isArray(val) && val !== null
    ) {
      result[key] = deepMerge(
        result[key] as Record<string, unknown>,
        val as Record<string, unknown>
      );
    } else {
      result[key] = val;
    }
  }
  return result;
}

function resolvePattern(pattern: string, ctx: Record<string, unknown>): string {
  let result = pattern.replace(/\{(\w+)\}/g, (_, key) => String(ctx[key] ?? ""));
  result = result.replace(/[-_\s]{2,}/g, (m) => m[0]);
  return result.replace(/^[-_ ]+|[-_ ]+$/g, "");
}

function isLangMap(val: unknown): val is Record<string, string> {
  return (
    typeof val === "object" && val !== null && !Array.isArray(val) &&
    Object.keys(val as object).every((k) => k.length === 2)
  );
}

function applyDerived(
  target: Record<string, unknown>,
  derived: Record<string, unknown>,
  ctx: Record<string, unknown>
): void {
  for (const [key, val] of Object.entries(derived)) {
    if (isLangMap(val)) {
      target[key] = Object.entries(val).map(([lang, pattern]) => ({
        language: lang,
        text: resolvePattern(pattern, ctx),
      }));
    } else if (typeof val === "object" && val !== null && !Array.isArray(val)) {
      const child = (target[key] as Record<string, unknown>) ?? {};
      applyDerived(child, val as Record<string, unknown>, ctx);
      target[key] = child;
    } else if (typeof val === "string") {
      target[key] = resolvePattern(val, ctx);
    }
  }
}

export function resolvePresetType(doc: Record<string, unknown>): Record<string, unknown> {
  const typeName = doc.type as string | undefined;
  if (!typeName) return doc;
  const typePath = path.join(getGeneratorPath(), "shell_templates", `${typeName}.yaml`);
  const raw = fs.readFileSync(typePath, "utf-8");
  const typeShell = yaml.load(raw) as Record<string, unknown>;
  const { type: _, ...rest } = doc;
  const merged = deepMerge(typeShell, rest);
  merged.shell = typeShell.shell;
  const derived = typeShell.derived as Record<string, unknown> | undefined;
  if (derived) {
    const matProps =
      ((merged.submodels as Record<string, unknown>)?.Properties as Record<string, unknown>)
        ?.MaterialProperties as Record<string, unknown> ?? {};
    const submodels = (merged.submodels as Record<string, unknown>) ?? {};
    applyDerived(submodels, derived, matProps);
    merged.submodels = submodels;
  }
  delete merged.derived;
  return merged;
}

File: Implementation2.0/ClassesAndBuilderMethods/BaSyx_AAS_Generator/shell_type_utils.py
import re
import yaml
from pathlib import Path

TYPE_SHELLS_DIR = Path(__file__).parent / "shell_templates"


def _deep_merge(base: dict, override: dict) -> dict:
    result = dict(base)
    for key, val in override.items():
        if key in result and isinstance(result[key], dict) and isinstance(val, dict):
            result[key] = _deep_merge(result[key], val)
        else:
            result[key] = val
    return result


def _build_token_context(merged: dict) -> dict:
    ctx = {}
    mat_props = merged.get("submodels", {}).get("Properties", {}).get("MaterialProperties", {})
    ctx.update(mat_props)
    return ctx


def _resolve_pattern(pattern: str, ctx: dict) -> str:
    def replace(m):
        return str(ctx.get(m.group(1), ""))
    result = re.sub(r"\{(\w+)\}", replace, pattern)
    result = re.sub(r"[-_\s]{2,}", lambda m: m.group(0)[0], result)
    return result.strip("-_ ")


def _is_lang_map(val: dict) -> bool:
    return bool(val) and all(
        isinstance(k, str) and len(k) == 2 and isinstance(v, str)
        for k, v in val.items()
    )


def _apply_derived(target: dict, derived: dict, ctx: dict) -> None:
    for key, val in derived.items():
        if isinstance(val, dict):
            if _is_lang_map(val):
                target[key] = [
                    {"language": lang, "text": _resolve_pattern(pattern, ctx)}
                    for lang, pattern in val.items()
                ]
            else:
                _apply_derived(target.setdefault(key, {}), val, ctx)
        elif isinstance(val, str):
            target[key] = _resolve_pattern(val, ctx)


def resolve_type(preset: dict) -> dict:
    """
    If preset has 'type', load the type shell, merge preset delta on top,
    then resolve any 'derived' patterns using MaterialProperties as token context.
    Presets without 'type' (using 'shell' directly) are returned unchanged.
    """
    type_name = preset.get("type")
    if not type_name:
        return preset
    path = TYPE_SHELLS_DIR / f"{type_name}.yaml"
    try:
        with open(path, encoding="utf-8") as f:
            type_shell = yaml.safe_load(f)
    except FileNotFoundError:
        raise FileNotFoundError(
            f"Type shell '{type_name}' not found at {path}. "
            f"Check the 'type:' field in your preset."
        )
    merged = _deep_merge(type_shell, {k: v for k, v in preset.items() if k != "type"})
    merged["shell"] = type_shell["shell"]
    derived = type_shell.get("derived", {})
    if derived:
        ctx = _build_token_context(merged)
        _apply_derived(merged.setdefault("submodels", {}), derived, ctx)
    merged.pop("derived", None)
    return merged
